mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoUpdateMethodGenerator.java

```diff
@@ -0,0 +1,196 @@
+
+  private static final EnumSet<ReturnTypeKind> SUPPORTED_RETURN_TYPES =
+      EnumSet.of(
+          VOID, FUTURE_OF_VOID, RESULT_SET, FUTURE_OF_ASYNC_RESULT_SET, BOOLEAN, FUTURE_OF_BOOLEAN);
+
+  public DaoUpdateMethodGenerator(
+      ExecutableElement methodElement,
+      DaoImplementationSharedCode enclosingClass,
+      ProcessorContext context) {
+    super(methodElement, enclosingClass, context);
+  }
+
+  @Override
+  public Optional<MethodSpec> generate() {
+
+    // Validate the parameters:
+    // - the first one must be the entity.
+    // - the others are completely free-form (they'll be used as additional bind variables)
+    if (methodElement.getParameters().isEmpty()) {
+      context
+          .getMessager()
+          .error(
+              methodElement,
+              "Wrong number of parameters: %s methods must have at least one",
+              Update.class.getSimpleName());
+      return Optional.empty();
+    }
+    VariableElement firstParameter = methodElement.getParameters().get(0);
+    TypeElement entityElement = asEntityElement(firstParameter);
+    if (entityElement == null) {
+      context
+          .getMessager()
+          .error(
+              methodElement,
+              "Invalid parameter type: "
+                  + "%s methods must take the entity to update as the first parameter",
+              Update.class.getSimpleName());
+      return Optional.empty();
+    }
+
+    // Validate the return type:
+    ReturnType returnType = parseReturnType(methodElement.getReturnType());
+    if (!SUPPORTED_RETURN_TYPES.contains(returnType.kind)) {
+      context
+          .getMessager()
+          .error(
+              methodElement,
+              "Invalid return type: %s methods must return either void, ResultSet or boolean "
+                  + "(possibly wrapped in a CompletionStage/CompletableFuture)",
+              Update.class.getSimpleName());
+      return Optional.empty();
+    }
+
+    // Generate the method:
+    String helperFieldName = enclosingClass.addEntityHelperField(ClassName.get(entityElement));
+    String statementName =
+        enclosingClass.addPreparedStatement(
+            methodElement,
+            (methodBuilder, requestName) ->
+                generatePrepareRequest(methodBuilder, requestName, helperFieldName));
+
+    MethodSpec.Builder methodBuilder = GeneratedCodePatterns.override(methodElement);
+
+    if (returnType.kind.isAsync) {
+      methodBuilder.beginControlFlow("try");
+    }
+    methodBuilder.addStatement(
+        "$T boundStatementBuilder = $L.boundStatementBuilder()",
+        BoundStatementBuilder.class,
+        statementName);
+
+    List<? extends VariableElement> parameters = methodElement.getParameters();
+    String entityParameterName = parameters.get(0).getSimpleName().toString();
+    methodBuilder.addStatement(
+        "$L.set($L, boundStatementBuilder)", helperFieldName, entityParameterName);
```

Comment:
There is a problem here: if the annotation has a custom WHERE clause, we can't set the whole entity at once because there might be no placeholders for PK components.

You can observe it by changing the DAO in the integration test like this:
```java
    @Update(customWhereClause = "id IN (:id1, :id2)")
    void updateWhereIdIn(Product product, UUID id1, UUID id2);
```
(currently it uses `id IN (:id, :id2)`, which masks the issue)

So you need to add a conditional here:
* if `customWhereClause` is empty: OK, do what you are doing right now
* else: you need to bind only the *non-PK* properties of the entity. You can copy what the Delete code does [here](https://github.com/datastax/java-driver/blob/java2078/mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoDeleteMethodGenerator.java#L171), it's almost the same except that you need to traverse `getRegularColumns` instead of `getPrimaryKey`.

Reply:
I think that this situation is not possible. Please take a look at a test in this commit:
https://github.com/datastax/java-driver/pull/1235/commits/74251c9ea3fc7d02fe45be6eccfa984b9c36c34a
1. I created the table with PK and Clustering: https://github.com/datastax/java-driver/pull/1235/commits/74251c9ea3fc7d02fe45be6eccfa984b9c36c34a#diff-8337f2ae926b239e7dcfbcb474e854ffR36
2. I've created a DAO that has no PK nor clustering: 
https://github.com/datastax/java-driver/pull/1235/commits/74251c9ea3fc7d02fe45be6eccfa984b9c36c34a#diff-8337f2ae926b239e7dcfbcb474e854ffR124
3. Then there is an update with partialDao: 
https://github.com/datastax/java-driver/pull/1235/commits/74251c9ea3fc7d02fe45be6eccfa984b9c36c34a#diff-7ab34dc135c21c80962589033fdf5217R484
Please note the `customWhereClause`:  `@Update(customWhereClause = "id IN (:id, :id2) AND clustering = 1")`

If you will remove the `AND clustering = 1` part you will get the server side fail:
`com.datastax.oss.driver.api.core.servererrors.InvalidQueryException: Some clustering keys are missing: clustering`
So it means that the user cannot create a where clause that doesn't have all PKs(including Clustering Keys) in it.

The update is setting regular columns only: https://github.com/datastax/java-driver/pull/1235/files#diff-b2adc444056441bbb59c5f7f513ce380R70 already so there should be no problems.

Please let me know if I am missing something.

Reply:
> id IN (:id, :id2) 

This only works because you named your placeholder the same as the column. Even if all PK fields must be present, you can't be sure that the user will use the correct names (as opposed to generated clauses where the processor controls the names).

I pushed the example from my previous comment. The query is valid, but if you run the test you will see that it fails because the processor is looking for a placeholder that isn't here:
```
java.lang.IllegalArgumentException: id is not a variable in this bound statement
```

Reply:
Ok, thx I didn't catch that from the previous comment (I thought that this is a matter of DTO).
It is fixed now